pkg/gatewayserver/gatewayserver.go

```diff
@@ -331,11 +358,21 @@ func (gs *GatewayServer) Connect(ctx context.Context, frontend io.Frontend, ids
 		return nil, err
 	}
 
-	conn := io.NewConnection(ctx, frontend.Protocol(), gtw, fp, scheduler)
+	conn := io.NewConnection(ctx, cancelCtx, frontend.Protocol(), gtw, fp, scheduler)
```

Comment:
So after this call, if you want to use the connection context, just take `conn.Context()`, or even re-assign the local `ctx`